src/crypto/crypto_kdf.cc

```diff
@@ -0,0 +1,131 @@
+
+  Utf8Value algorithm(env->isolate(), args[offset]);
+  params->kdfCtx = ncrypto::KdfCtxPointer::FromName(algorithm.ToStringView());
+
+  // if (!pass.CheckSizeInt32()) [[unlikely]] {
+  //   THROW_ERR_OUT_OF_RANGE(env, "pass is too large");
+  //   return v8::Nothing<void>();
+  // }
+  // if (!salt.CheckSizeInt32()) [[unlikely]] {
+  //   THROW_ERR_OUT_OF_RANGE(env, "salt is too large");
+  //   return v8::Nothing<void>();
+  // }
+
+  auto paramsObj = args[offset + 1].As<v8::Object>();
+  auto propertyNames = paramsObj->GetOwnPropertyNames(env->context()).ToLocalChecked();
+
+  for (uint32_t i = 0; i < propertyNames->Length(); ++i) {
+    auto key = propertyNames->Get(env->context(), i).ToLocalChecked();
+    if (!key->IsString()) {
+      continue;
+    }
+
+    auto name = Utf8Value(env->isolate(), key).ToString();
+    auto value = paramsObj->Get(env->context(), key).ToLocalChecked();
+
+    if (value->IsUint32()) {
+      params->kdfParams[name] = value.As<v8::Uint32>()->Value();
+    } else if (value->IsNumber()) {
+      params->kdfParams[name] = value.As<v8::Number>()->Value();
+    } else if (value->IsString()) {
+      params->kdfParams[name] = Utf8Value(env->isolate(), value).ToString();
+    } else if (IsAnyBufferSource(value)) {
+      ArrayBufferOrViewContents<char> abv(value);
+      params->kdfParams[name] = std::vector<char>(abv.data(), abv.data() + abv.size());
+    }
+  }
+
+  if (!params->kdfCtx.setParams(params->kdfParams)) {
+    if (uint32_t err = ERR_peek_last_error(); err != 0) {
+      THROW_ERR_CRYPTO_INVALID_KDF_PARAMS(
+          env, "Invalid kdf params: %s", ERR_error_string(err, nullptr));
+    } else {
+      THROW_ERR_CRYPTO_INVALID_KDF_PARAMS(env);
+    }
+    return v8::Nothing<void>();
+  }
+  
+  params->length = args[offset + 2].As<v8::Uint32>()->Value();
+  if (auto kdfSize = params->kdfCtx.getSize(); kdfSize == 0) {
+    if (uint32_t err = ERR_peek_last_error(); err != 0) {
+      THROW_ERR_CRYPTO_INVALID_KDF_PARAMS(
+          env, "Invalid kdf params: %s", ERR_error_string(err, nullptr));
+    } else {
+      THROW_ERR_CRYPTO_INVALID_KDF_PARAMS(env);
+    }
+    return v8::Nothing<void>();
+  } else if (kdfSize != SIZE_MAX && params->length != kdfSize) {
+    THROW_ERR_CRYPTO_INVALID_KEYLEN(env);
+    return v8::Nothing<void>();
+  }
+
+  return v8::JustVoid();
+}
+
+bool KDFTraits::DeriveBits(
+    Environment* env,
+    const KDFConfig& params,
+    ByteSource* out) {
+  // If the params.length is zero-length, just return an empty buffer.
+  // It's useless, yes, but allowed via the API.
+  if (params.length == 0) {
+    *out = {};
+    return true;
+  }
+
+  auto dp = params.kdfCtx.derive(params.length);
+
+  if (!dp) return false;
+  DCHECK(!dp.isSecure());
+  *out = ByteSource::Allocated(dp.release());
+  return true;
+}
+
+// helper type for the visitor
+template<class... Ts>
+struct overloads : Ts... { using Ts::operator()...; };
+
+void KDFConfig::MemoryInfo(MemoryTracker* tracker) const {
```

Comment:
Take a look at `MemoryRetainerTraits` ... it was recently added to simplify tracking for types that cannot be modified to implement `MemoryRetainer` https://github.com/nodejs/node/blob/main/src/memory_tracker.h#L141-L166